File: leetcode/src/python3/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # 两个节点：
    # fast 和 slow 相距 n 个节点，当 fast 到达链表结尾时，slow 就刚好在倒数第 n 个节点上.
    # 为了方便 slow 好删除节点，刚开始 fast 和 slow 要相距 n + 1 个节点。
    # 步骤：
    # 1. 先让 fast 从初始位置走 n + 1 步
    # 2. 让 slow 和 fast 同时向前走, fast 走到链表末尾停止
    # 3. slow 指向倒数第 n + 1 个节点，进行删除操作
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        # 让 fast 比 slow 先走一步
        fast = head
        slow = dummy

        # fast 再走 n 步
        # fast 和 slow 相差 n + 1 步
        for i in range(n):
            fast = fast.next

        while fast:
            fast = fast.next
            slow = slow.next
        
        # 删除节点
        slow.next = slow.next.next
        return dummy.next
# @lc code=end

File: leetcode/src/rust/19.remove-nth-node-from-end-of-list.rs
/*
 * @lc app=leetcode id=19 lang=rust
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        // -1 -> 1 -> 2 -> 3 -> 4 -> 5
        // ^^^   ^^^
        // dummy head
        let mut dummy = Box::new(ListNode {
            val: -1,
            next: head,
        });
        // -1 -> 1 -> 2 -> 3 -> 4 -> 5
        // ^^^ 
        // right  and left
        let mut right = dummy.clone();
        let mut left = dummy.as_mut();

        // -1 -> 1 -> 2 -> 3 -> 4 -> 5
        // ^^^       ^^^ 
        // left     right
        for _ in 0..n {
            right = right.next.unwrap();
        }

        // -1 -> 1 -> 2 -> 3 -> 4 -> 5
        //                ^^^       ^^^ 
        //                left     right
        while let Some(node) = right.next {
            right = node;
            left = left.next.as_mut().unwrap();
        }
        
        // 删除倒数第二个节点
        left.next = left.next.as_mut().unwrap().next.clone();

        dummy.next
    }
}
// @lc code=end

